### PR TITLE
For #26424 - Create wallpaper file migration helper

### LINKED ISSUE #26424: Create wallpaper file migration helper
We plan to simplify the number of assets possible for each wallpaper. Previously, there were combinations of all of portrait/landscape, light/dark, and high/medium/low resolutions for each wallpaper. We will no longer be providing separate assets for themes or resolutions, so there will only be separate assets for different orientation (and thumbnails).

This will mean that currently saved versions of wallpapers will be stored under directory paths that are no longer relevant, like `wallpapers/<orientation>/<theme>/<wallpaperName>.png`. The new paths should follow a convention like `wallpapers/<wallpaperName>/portrait.png` and `wallpapers/<wallpaperName>/landscape.png`.

This will require migrating existing assets to the new path names. This is required because some users may have expired wallpapers that are no longer available, so we cannot just download a fresh asset. It is likely fine to ignore the consideration of updating the themed wallpaper to a non-themed wallpaper. Both the landscape and the portrait versions of an image should be copied.

If there is not a thumbnail available for the wallpaper (as could be the case with Turning Red wallpapers that are no longer available), then the portrait image of the wallpaper should also be copied as `wallpapers/<wallpaperName>/thumbnail.png` to ensure that it can be visible in the wallpaper settings page.

AC:
1. Install a build of the app that is not using the `wallpapersV2Enabled` feature flag, then set a wallpaper.
2. Overwrite it with a build that includes the migration code and has the flag enabled.
3. The selected wallpaper should still be selected.
4. The selected wallpaper can be rotated correctly.
5. The wallpaper is visible in the settings page. For wallpapers that have a thumbnail available remotely, it should use the new remote version.

The `WallpaperDownloader` and `WallpaperFileManager` contain code that would be a good starting point to building context around this issue.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21267)

### PR DESCRIPTION
**Pull request by Alexandru2909**
Migrate legacy wallpapers to the respective V2 paths. If a thumbnail for the legacy wallpaper is available remote, then it should be used, otherwise the portrait wallpapers should be used instead.

### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #26424

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperDownloader.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperDownloader.kt:74]: will this conflict with https://github.com/mozilla-mobile/fenix/pull/26562?

Comment (Alexandru2909) [reply app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperDownloader.kt:74]: Yes, I'll rebase and remove this function once the work for #26562 is merged. Since we'll be waiting for the PR to land I'll add the do-not-land label here.

File: app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt:233]: nit: could you add the named parameters here

File: app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt:234]: will the migration code be ran every time this use case is invoked?

File: app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperFileManager.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperFileManager.kt:142]: does there need to be any File exception handling for any of the copy/delete operations?

File: app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt:272]: Including `migratedWallpapers` here shouldn't be necessary, as there is not currently a requirement to display an expired wallpaper in the available wallpaper list in the settings page.

File: app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperFileManager.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperFileManager.kt:59]: We should check with product before removing this `clean` method. It is intended to remove promotional wallpaper files once those promotions expire. During the Turning Red promotion, there was quite a bit of focus on making sure users would not be able to access Turning Red assets that they had not intended to keep. I would be concerned that we might have similar requirements from promotional partners in the future.

File: app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperFileManager.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperFileManager.kt:64]: It might make sense to put the migration code into a specialized `WallpaperLegacyMigration` class, so that the whole thing can be easily deleted once we're satisfied that enough time has passed.

File: app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperFileManager.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperFileManager.kt:141]: I applaud the thoroughness of downloading a correct thumbnail if it exists. In the interest of simplicity, I think it would also be fine to just use the portrait image as the thumbnail. That would also ensure the thumbnail matches the legacy version of the wallpaper.

**Review by MatthewTighe — COMMENTED** [review]
This is another case of me needing to specify more issues more clearly. The main goal of this issue is to ensure that users don't lose access to their selected wallpaper, or need to re-select it when they update to 106. Because of that, I don't think we need to migrate all existing assets that a user has. We should be able to migrate only their currently selected wallpaper.

File: app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt
Comment (Alexandru2909) [reply app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt]: Just to be sure this is the intended behavior: When testing this PR I noticed the migrated wallpaper (while still set) is not displayed in the wallpaper settings. Therefore if the user selects any other wallpaper the migrated one will be cleaned on the next startup. Here is a video of the behavior I described: 

https://user-images.githubusercontent.com/35462038/188904518-938e5480-7d96-407d-a223-7b63335cdb3b.mp4

Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt]: Yes, that behavior is intended (if suboptimal). During wallpapers V1 it was decided that users could keep their selected wallpaper once a promotion ended, but it would no longer appear in settings and if the wallpaper was switched the user would lose access to it. This was mostly a "this is the easiest thing to do for now" decision.

There was some talk about improving that in the future, such that once a user selected a wallpaper it would be available in their "collection" indefinitely. However, such an improvement I think is outside the scope of this issue and could instead be done during further iterations of the feature.

Totally unrelated: I spotted your deftones search - nice taste lol

File: app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt:249]: Rather than include this here, I think it would make sense to run the migration separately after the current wallpaper is dispatched on line 240. We could also wrap the call in a conditional based on a property in `Settings` so that we only do it once. I believe I saw that included in a previous version of this patch. Something like:

```
Wallpaper.getCurrentWallpaperFromSettings(settings)?.let {
    store.dispatch(AppAction.WallpaperAction.UpdateCurrentWallpaper(it))
}
val currentWallpaperName = withContext(Dispatchers.IO) { settings.currentWallpaperName }
if (settings.requiresWallpaperMigration) {
    migrationHelper.migrateLegacyWallpaper(currentWallpaperName)
}
```

Once the wallpaper is migrated, it should get picked up automatically as part of `possibleWallpapers` or `lookupExpiredWallpaper`. 

My thinking is that this change will make it easier to remove the migration code in the future. Also, it should make it more clear that the migration code is only running once, rather than relying on the conditional in the migration code that looks like `if (!legacyLandscapeFile.exists() || !legacyPortraitFile.exists())`. This will also skip the IO of looking for the legacy files.

**Review by MatthewTighe — APPROVED** [review]
One more small comment, but this is otherwise looking good to me! I'll approve now so it can be landed during your next day if you'd like

**Comment by MatthewTighe** [comment]
:shipit:

File: app/src/main/java/org/mozilla/fenix/wallpapers/LegacyWallpaperMigration.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/wallpapers/LegacyWallpaperMigration.kt:70]: There is one more special case that I would like to cover here if you don't mind:

If the wallpaper has been successfully migrated then:

```
if (wallpaperName == "mei" || wallpaperName == "panda") {
    settings.currentWallpaperTextColor = "FFFBFBFE".toLong(radix = 16)
}
```

this will allow us to use an appropriate text color if a user is still using one of the expired Turning Red wallpapers.

**Review by MatthewTighe — APPROVED** [review]
One last minute comment based on UX feedback. This is otherwise blocking https://github.com/mozilla-mobile/fenix/pull/26967 so please land it soon if you are feeling comfortable with it

**Comment by Mugurell** [comment]
Can help with getting this landable while Alex is in PTO.
- addressing https://github.com/mozilla-mobile/fenix/pull/26685#discussion_r970088593
- adding trailing commas

**Comment by Mugurell** [comment]
Added the trailing command in the first commit, and support for the migrated Turning Red wallpapers in the second commit.

**Pull request merged**
